File: swea/ssafy/24444.py
from collections import deque
from sys import stdin, stdout

def bfs():
    cnt = 1
    q = deque()
    q.append(r)
    visit[r] = cnt
    cnt += 1
    while q:
        v = q.popleft()
        for i in lst[v]:
            if not visit[i]:
                visit[i] = cnt
                cnt += 1
                q.append(i)


n, m, r = map(int, stdin.readline().split())
lst = [[] for _ in range(n+1)]
visit = [0 for _ in range(n+1)]
for i in range(m):
    a, b = map(int, stdin.readline().split())
    lst[a].append(b)
    lst[b].append(a)
for i in range(1, n+1):
    lst[i].sort()
bfs()
for i in range(1, n+1):
    stdout.write(str(visit[i])+'\n')

File: c/Project1/Project1/24444.cpp
/*
#include <iostream>
#include <deque>
#include <vector>
#include <algorithm>
using namespace std;
//int** lst = new int*[100001];
int* visit = new int[100001]{};
vector<int> e[100001];
int n, m, r;
void bfs(int);

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	int a, b;
	cin >> n >> m >> r;
	for (int i = 0; i < m; i++)
	{
		cin >> a >> b;
		e[a].push_back(b);
		e[b].push_back(a);
	}
	for (int i = 1; i < n + 1; i++)
	{
		sort(e[i].begin(), e[i].end());
	}
	bfs(r);
	cout << visit[1];
	for (int i = 2; i < n + 1; i++)
	{
		cout << '\n' << visit[i];
	}
	return 0;
}

void bfs(int r) {
	int cnt = 0;
	deque<int> q;
	q.push_back(r);
	visit[r] = ++cnt;
	while (!q.empty())
	{
		int v = q.front();
		q.pop_front();
		for (int i : e[v])
		{
			if (visit[i] == 0)
			{
				visit[i] = ++cnt;
				q.push_back(i);
			}
		}
	}
}
*/